torch/distributed/optim/post_localSGD_optimizer.py

```diff
@@ -87,7 +78,11 @@ def step(self):
         Performs a single optimization step (parameter update).
         """
         self.optim.step()
-        self.averager.average_parameters(iter(self.params))
+        for param_group in self.param_groups:
+            for params in param_group["params"]:
+                if params.grad is None:
+                    continue
+                self.averager.average_parameters(iter(params))
```

Comment:
this is before this PR, but curious if we want to support parameter averaging in an async way? We could return futures here and then set up some scaffolding to wait on these parameters before they are used in forward pass. Or is this approach flawed?

Reply:
That's actually a good idea! We can definitely do it in an async way, in combination with a built-in `forward_pre_hook`.

Again, it depends on whether if there is such a requirement, or how much extra performance gain we can achieve, which also depends on model averaging frequency.

Currently in my use case model averaging is not a performance bottleneck, so I would prefer supporting this feature in the future on demand, or as a bootcamp task. After all, the downside of supporting another async way may complicate the API. We need to find a use case or requirement to justify such complication first.

Reply:
Oh yeah, forgot that frequency is only something like once every 16 iters, so maybe overlapping perf improvements aren't the most pressing right now.